### ISSUE TITLE
cuelang.org: preprocessor should ensure all published CUE is cue-fix'd

### ISSUE THREAD

**Issue by jpluscplusm** [cue-lang/cue#3419]
### What page were you looking at?

https://cuelang.org/

### What version of the site were you looking at?

https://github.com/cue-lang/cuelang.org/commit/6ccc4f09149fb0487cd433a45332714d3286ae98

### What did you expect?

I expected that all CUE presented on the site would be idiomatic, in terms of both syntax and formatting.

We currently ensure idiomatic formatting by passing CUE through `cue fmt`, in the preprocessor.

We should first (instead?) pass the CUE through `cue fix` to ensure that the site demonstrates idiomatic CUE, as determined by the version of CUE actually presented on the site. I opened #3417 to request that `cue fix` gain the ability to communicate via stdin/stdout, as this capability gap is the reason that I didn't simply open a speculative CL making this change (swapping `cue fmt` for `cue fix`).

### What did you see instead?

It's possible to author pages relying on outdated CUE syntax which will never be automatically caught or flagged until some future version of CUE stops supporting it entirely and the site is upgraded to that version - complicating the site's upgrade (which should be simple), not to mention implying that outdated CUE would have been presented on the site longer than it should have been.

**Comment by myitcv** [comment]
cc @mvdan @cuematthew 

I was perhaps falsely under the impression that `cue fmt` applied the same fixes as `cue fix`?

**Comment by cuematthew** [comment]
As far as I can see, the only code paths that lead to fixing come from the `cue fix` command.

**Comment by mvdan** [comment]
Hm. What would the point be of having `cue fmt` apply fixes if we also have `cue fix`? Wouldn't `cue fix` be redundant then?

It seems like you're sort of right - `cue fmt` did apply fixes in the past. I believe I unintentionally removed this in a recent refactor in https://review.gerrithub.io/c/cue-lang/cue/+/1194169 - note the call to `fix.File`. We did not notice because, unsurprisingly, there were zero tests of this feature.

We can bring it back with tests for sure, but then I'm confused why we need both `cue fmt` and `cue fix`. My impression is that we should leave `cue fmt` to only formatting, and any semantic rewrites (for example to move away from deprecated APIs or removed language features) to `cue fix`, akin to Go.

If, alternatively, we want `cue fmt` to regain its fixing ability - then I would imagine we could delete `cue fix` entirely.

**Comment by mvdan** [comment]
Summary of the discussion on Discord:
* `cue fmt` was indeed meant to apply a subset of fixes which were needed very soon, and were 100% correct.
* This was for the sake of convenience; most users did not run `cue fix` regularly.
* It seems like such a subset was never consistently implemented, documented, or tested; for example, https://pkg.go.dev/cuelang.org/go/tools/fix@master does not have this separation, and my change to `cue fmt` above did not break any tests.
* Matthew and I prefer `cue fmt` to remain just about formatting, and not any more significant fixes or rewrites. Paul seems to have a similar opinion for the sake of the LSP.

We are then happy to keep the two commands as they are on master: `cue fmt` only does formatting, and `cue fix` does formatting plus all fixes. We will lean into `cue fix` and `cue mod fix`, mentioning them in upcoming release notes and error messages.

Note that `cue fix` MUST currently do formatting too, because `cue/format` is the only way to write CUE back to disk, and it formats the CUE too. This seems reasonable to me; my usual workflow with a new CUE version would be to first `cue fmt ./...`, commit those very safe formatting changes, and then run `cue fix ./...` and do a second commit, carefully reviewing that all fixed code still looks and behaves correctly.

TL;DR the website will indeed need to run `cue fix` too, if it wants fixes. I already sent a fix for https://github.com/cue-lang/cue/issues/3417 too.

**Comment by myitcv** [comment]
Thanks for the great summary, @mvdan. 

@jpluscplusm - we will keep this issue open to track the adding of support for running `cue fix` instead of `cue fmt`, unless a block specifies the `#nofix` directive or similar.

**Comment by ofek** [comment]
Has this been completed?
